Add language reference material for `unsafe` updates for memory safety (#54367)
dotnet/docs: docs/csharp/language-reference/unsafe-code.md

[!INCLUDE[csharp-version-note](./includes/initial-version.md)]

> C# also supports an [`unsafe`](keywords/unsafe.md) context, in which you can write *unverifiable* code. Unsafe code isn't necessarily dangerous; it's just code whose safety .NET tools can't verify. You use unsafe code to call native functions that require pointers, and in some cases to improve performance through direct memory access that avoids array bounds checks. Unsafe code also introduces security and stability risks. To compile code that contains an `unsafe` context, add the [**AllowUnsafeBlocks**](compiler-options/language.md#allowunsafeblocks) compiler option.

Reviewer: ```suggestion
C# also supports an [`unsafe`](keywords/unsafe.md) context, in which you can write *unverifiable* code. Unsafe code isn't necessarily dangerous; it's just code whose safety can't be verified by .NET tools. You use unsafe code to call native functions that require pointers, and in some cases to improve performance through direct memory access that avoids array-bounds checks. Unsafe code also introduces security and stability risks. To compile code that contains an `unsafe` context, add the [**AllowUnsafeBlocks**](compiler-options/language.md#allowunsafeblocks) compiler option.
```

> C# defines two models for what counts as unsafe code: the original model and an updated memory safety model that's in preview in C# 15 and .NET 11. For how the two models differ, see [Two models for unsafe code](#two-models-for-unsafe-code).

Reviewer: ```suggestion
C# defines two models for what counts as unsafe code: the original model and an updated memory safety model that's in preview in C# 15 and .NET 11. For information about how the two models differ, see [Two models for unsafe code](#two-models-for-unsafe-code).
```

You can learn more about function pointers in the [Function pointers](~/_csharpstandard/standard/unsafe-code.md#2433-function-pointers) section of the C# language specification.

### Example: Use pointers to copy an array of bytes

The following example uses pointers to copy bytes from one array to another.

> This example uses the [unsafe](keywords/unsafe.md) keyword, which enables you to use pointers in the `Copy` method. The [fixed](statements/fixed.md) statement declares pointers to the source and destination arrays. The `fixed` statement *pins* the location of the source and destination arrays in memory so that garbage collection doesn't move the arrays. The `fixed` block pins the memory blocks for the arrays in the scope of the block. Because the `Copy` method in this example uses the `unsafe` keyword, you must compile it by using the [**AllowUnsafeBlocks**](compiler-options/language.md#allowunsafeblocks) compiler option.

Reviewer: ```suggestion
This example uses the [`unsafe`](keywords/unsafe.md) keyword, which enables you to use pointers in the `Copy` method. The [`fixed`](statements/fixed.md) statement declares pointers to the source and destination arrays. The `fixed` statement *pins* the location of the source and destination arrays in memory so that garbage collection doesn't move the arrays. The `fixed` block pins the memory blocks for the arrays in the scope of the block. Because the `Copy` method in this example uses the `unsafe` keyword, you must compile it by using the [**AllowUnsafeBlocks**](compiler-options/language.md#allowunsafeblocks) compiler option.
```

    }
}
```

A caller wraps the call in its own `unsafe` block:

```csharp
// Preview
unsafe
{
    int value = ReadInt32(buffer);
}
```

The updated model also tightens a few related rules:

- The `unsafe` modifier produces an error on a type declaration, a static constructor, and a finalizer, because the modifier has no caller to inform.
- Delegates can't be `unsafe`, because a delegate is type-shaped.
- A type whose parameterless constructor is `unsafe` doesn't satisfy the `new()` constraint.

### Operations that require an unsafe context

Operations that access the pointed-to memory require an `unsafe` context:

- Pointer indirection (`*p`), pointer member access (`p->member`), and pointer element access (`p[i]`).
- Function pointer invocation.
- Element access on a fixed-size buffer.

The following example pins an array without an `unsafe` context but dereferences the pointer inside one:

:::code language="csharp" source="snippets/memory-safety/Relaxations.cs" id="Dereference":::

### Relaxed operations

Operations that don't access pointed-to memory no longer require an `unsafe` context:

- Declaring a pointer type and taking the address of a variable with the `&` operator.

> - The [`fixed`](statements/fixed.md) statement that pins a variable.

Reviewer: If the fixed statement can do things other than pin a variable, then ignore my previous comment on this same text that appears elsewhere.

Author: That's all it does.

### The safe keyword

The updated model adds a [`safe`](keywords/safe.md) contextual keyword that attests a declaration is sound where the compiler requires you to make the choice explicit.

An `extern` member calls into native code, so the compiler can't classify its safety. Under the updated model, you mark every `extern` declaration, including a `LibraryImport` partial method, either `safe` or `unsafe`:

```csharp
// Preview
[LibraryImport("libc")]
internal static safe partial int getpid();

[LibraryImport("libc", StringMarshalling = StringMarshalling.Utf8)]
internal static unsafe partial nint strlen(byte* str);
```

`getpid` takes no parameters and returns a primitive, so the author attests that the call is safe and callers use it without ceremony. `strlen` takes a raw pointer that the native code dereferences, so the declaration is `unsafe` and propagates the obligation to callers. Omitting both modifiers is an error, which forces you to make the safety decision. A field in a struct with explicit layout uses the same rule.

> ### Opt in and cross-assembly behavior

Reviewer: ```suggestion
### Opt-in and cross-assembly behavior
```
